proposals/0000-local-do.rst

```diff
@@ -0,0 +1,1058 @@
+expressions <https://docs.microsoft.com/en-us/dotnet/fsharp/language-reference/computation-express
+ions>`_.
+
+Motivation
+------------
+
+There are many kinds of monad-like things out there:
+
+* monads
+* indexed monads
+* graded monads
+* relative monads
+* linear variants of all the above once `linear types
+  <https://github.com/ghc-proposals/ghc-proposals/pull/111>`_ are
+  implemented
+
+  * in particular, the linear IO monad from the `Linear Haskell paper
+    <https://arxiv.org/abs/1710.09756>`_ is an example of a linear
+    graded monad
+* …
+
+All of these are theoretically compatible with the do-notation. And, in fact, really want the do-notation to work for them. After all, the entire existence of the do-notation can be ascribed to: “it's really annoying to program with ``(>>=)`` all the time”.
+
+The prescribed solution is to use ``-XRebindableSyntax``. But ``-XRebindableSyntax`` is a very blunt instrument:
+
+* It affects many syntactic constructs (numerical literals, the ``if then else`` syntax, … (see the `full list <https://downloads.haskell.org/~ghc/latest/docs/html/users_guide/glasgow_exts.html#extension-RebindableSyntax>`_)).
+* It implies ``-XNoImplicitPrelude``
+* It uses the same rebinding for all the do-expression in an entire file.
+
+You may not want all this. For instance, with linear types ``if then else`` syntax cannot be meaningfully rebound to a function. And there is no reason why a file wouldn't have a piece code referring to a monad, one to a graded monad, and one to a linear relative monad.
+
+Proposed Change Specification
+-----------------------------
+
+This proposal creates a new language extension ``-XQualifiedDo``.
+
+When ``-XQualifiedDo`` is activated, the syntax of the ``do`` notation is changed to
+
+::
+
+  <lexp> ⟶ [modid.]do
+
+where ``modid`` stands for some module name.
+
+The additional module name is called the qualifier of the do-expression.
+
+The semantics of ``do`` notation statements is given as follows (using
+``-XLambdaCase`` notation):
+
+* The ``x <- u`` statement uses ``(modid.>>=)``
+
+  ::
+
+    M.do { x <- u; stmts }  =  u M.>>= \x -> M.do { stmts }
+
+* The ``u`` statement uses ``modid.(>>)``
+
+  ::
+
+    M.do { u; stmts }  =  u M.>> M.do { stmts }
+
+* The a ``pat <- u`` statement uses ``M.fail`` for the failing case,
+  if such a case is needed
+
+  ::
+
+    M.do { pat <- u; stmts }  =  u M.>>= \case
+      { pat -> M.do { stmts }
+      ; _ -> M.fail "…"
+      }
+
+  If the pattern cannot fail, then we don't need to use ``M.fail``.
+
+  ::
+
+    M.do { pat <- u; stmts }  =  u M.>>= \case pat -> M.do { stmts }
+
+* ``-XApplicativeDo`` uses ``(M.<$>)``, ``(M.<*>)`` and ``M.join`` (this
+  assumes that the applicative-do grouping has been performed)
+
+  ::
+
+    M.do { (x1 <- u1 | … | xn <- un); return e }  =
+      (\x1 … xn -> e) M.<$> u1 M.<*> … M.<*> un
```

Comment:
Wait, here is a snag. We decided elsewhere that unqualified uses of monad operations are untouched. Yet this translation is “matching” on an unqualified `return`! Shouldn't this be
```suggestion
    M.do { (x1 <- u1 | … | xn <- un); M.return e }  =
      (\x1 … xn -> e) M.<$> u1 M.<*> … M.<*> un
```

If so, what does it mean to match on `M.do … M.return`? Does it have to be the same `modid`? Or is `M1.do … M2.return` allowed if `M2.return` resolves to `A.B.C.return` and `M1` is an alias for `A.B.C`?

Reply:
Requiring the same `modid` is more restrictive, make sure the desugaring is still syntactically driven, and probably works in most cases, and is what people should do anyways.

So I suggest we start with this and relax if it turns out to be the wrong choice.

Reply:
The return is not used in the desugaring. Should it be demanded to be in scope? If not required, what sense does it make to qualify the return? I think it would be rather confusing to ask the user to pass a function that is never called!

I'd argue that return here is used as a keyword, not a function that the user should qualify.

Reply:
But `return` is not a keyword… 

Let’s see what happens with `ApplicativeDo`. It works if `return` is “the right `return`”:
```
Prelude> :set -XApplicativeDo 
Prelude> foo a b = do { x <- a; return (b x) }
Prelude> :t foo
foo :: Functor f => f t -> (t -> b) -> f b
```
but not when `return` is not in scope or some other `return` than `Prelude.return`:

```
Prelude> import Prelude hiding (return)
Prelude> :set -XApplicativeDo 
Prelude> foo a b = do { x <- a; return (b x) }

<interactive>:3:24: error:
    Variable not in scope: return :: t1 -> m b
Prelude> let return = error "any type"
Prelude> foo a b = do { x <- a; return (b x) }
Prelude> :t foo
foo :: Monad m => m t1 -> (t1 -> t2) -> m b
```

So it seems prudent that even with `QualifiedDo`, `ApplicativeDo` only works if the “right” `return` is used.

Reply:
Ok. Let's ask a qualified return then. Thanks for elaborating the motivation.

Reply:
Qualified with what? :-)

Reply:
Ah, I see your chnage. That’d be the “the same M”. I sounds good to me.